### PR TITLE
MushroomDao tests + minor fixes to MushroomDaoIMpl

### PR DESCRIPTION
**Pull request by bencikpeter**

### CODE REVIEW

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/MushroomDaoTest.java [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.*;
+
+/**
+ * @author bencikpeter
+ */
+@ContextConfiguration(classes= PersistenceSampleApplicationContext.class)
+@TestExecutionListeners(TransactionalTestExecutionListener.class)
+@Transactional
```
Comment (bkompis) [inline persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/MushroomDaoTest.java]: could use AbstractTransactionalJUnit4SpringContextTests

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/MushroomDaoTest.java [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+    @Before
+    public void setUp(){
+        // create mushrooms
+        mushroom1 = createMushroom("toadstool", MushroomType.POISONOUS,"June","July");
+        mushroom2 = createMushroom("champignon", MushroomType.EDIBLE,"May","September");
+
+        // persist mushrooms
+
+        em.persist(mushroom1);
+        em.persist(mushroom2);
```
Comment (bkompis) [inline persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/MushroomDaoTest.java:54]: missing call to em.flush(). Changes in whether following tests pass are not apparent, but flush should be done. (Or am I mistaken?)

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/MushroomDaoTest.java [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+        mushroom.setIntervalOfOccurrence(beginMonth,endMonth);
+
+        return mushroom;
+    }
+
+    @Test
+    public void findById_validId() throws Exception {
+        Mushroom mushroom = mushroomDao.findById(mushroom1.getId());
+        assertThat(mushroom).isNotNull();
+        assertThat(mushroom).isEqualToComparingFieldByField(mushroom1);
```
Comment (bkompis) [inline persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/MushroomDaoTest.java]: Use .isEqualTo(Object other), we're only interested in business equivalence.

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/MushroomDaoTest.java [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+    }
+
+    @Test
+    public void findById_invalidId() throws Exception {
+        Mushroom mushroom = mushroomDao.findById(mushroom1.getId()+20);
```
Comment (bkompis) [inline persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/MushroomDaoTest.java]: relies on a particular ID generating strategy in the db (what if db generated id for mushroom2 was mushroom1.getId()+20?). Possible solutions:
* add all known ID's
* work on an empty database

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/MushroomDaoTest.java [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+    }
+
+    @Test
+    public void findById_nullId() throws Exception {
+        assertThatThrownBy(() -> mushroomDao.findById(null)).hasRootCauseExactlyInstanceOf(IllegalArgumentException.class);
+    }
+
+    @Test
+    public void create() throws Exception {
+        Mushroom mushroom = createMushroom("WierdMushroom", MushroomType.UNEDIBLE, "May", "July");
```
Comment (bkompis) [inline persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/MushroomDaoTest.java]: typo in mushroom name :)

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/MushroomDaoTest.java [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+    }
+
+    @Test
+    public void create_nullMushroomType() throws Exception {
+        Mushroom mushroom = createMushroom("WierdMushroom", null, "May", "July");
+
+        assertThatThrownBy(() -> mushroomDao.create(mushroom)).isInstanceOf(ConstraintViolationException.class);
+    }
+
+    //this string validation is not implemented yet - tests prepared for the future
```
Comment (bkompis) [inline persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/MushroomDaoTest.java:112]: nice planning

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/MushroomDaoTest.java [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+    public void create_nullEntity() throws Exception {
+
+        assertThatThrownBy(() -> mushroomDao.create(null)).hasRootCauseExactlyInstanceOf(IllegalArgumentException.class);
+    }
+
+    @Test
+    public void delete() throws Exception {
+
+        mushroomDao.delete(mushroom1);
+
```
Comment (bkompis) [inline persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/MushroomDaoTest.java]: delete must be flushed

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/MushroomDaoTest.java [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+    public void delete_nullEntity() throws Exception {
+        assertThatThrownBy(() -> mushroomDao.delete(null)).hasRootCauseInstanceOf(IllegalArgumentException.class);
+    }
+
+    @Test
+    public void findAll() throws Exception {
+
+        List<Mushroom> list = mushroomDao.findAll();
+
+        assertThat(list).hasSize(2);
```
Comment (bkompis) [inline persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/MushroomDaoTest.java]: feel free to check the contents of the list too (containsExactlyInAnyOrder(...))

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/MushroomDaoTest.java [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+    @Test
+    public void findByMushroomType() throws Exception {
+        List<Mushroom> list = mushroomDao.findByMushroomType(MushroomType.EDIBLE);
+        assertThat(list).hasSize(1);
+    }
+
+    @Test
+    public void findByMushroomType_empty() throws Exception {
+        List<Mushroom> list = mushroomDao.findByMushroomType(MushroomType.UNEDIBLE);
+        assertThat(list).hasSize(0);
```
Comment (bkompis) [inline persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/MushroomDaoTest.java]: assertThat(list).isEmpty() :)

**Review by bkompis — CHANGES_REQUESTED** [review]
* use business equivalence to compare objects (not field by field)
* flush after em.persist() and dao update & delete
* check contents of containers

**Comment by bencikpeter** [comment]
All change requests fixed, some of them are just not accurately placed so they did not dissapear.

File: persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/MushroomDaoTest.java (hunk 1 above)
Comment (bkompis) [reply persistence/src/test/java/cz/muni/fi/pa165/mushrooms/dao/MushroomDaoTest.java]: @Transactional is now redundant

**Review by bkompis — APPROVED** [review]
Remove @Transactional on your test class, then I'll merge.

**Pull request merged**